enable CloudTrail action

skurfer opened the pull request:
Another stab at #339. This will enable the trail(s) named in the policy, or create then enable them (and a dedicated bucket) if they don’t exist.

kapilt commented:
actions now require a permissions tuple describing the permissions they need.

kapilt commented on c7n/resources/cloudtrail.py:
```diff
@@ -31,3 +40,93 @@ class resource_type(object):
+            policy_json = policy_json_tmpl.format(
+                bucket_name=bucket_name, account_id=account_id,
+            )
+            s3client.put_bucket_policy(Bucket=bucket_name, Policy=policy_json)
```
this needs to play nice with an existing bucket policy

skurfer replied:
The bucket was created one line earlier, so existing policy shouldn’t be a concern. Is it possible to define a default policy for newly created buckets? That’s the only way I could see it already having one.

kapilt replied:
thats fair, but the consideration is that we'll be doing this to trails back into place if their deleted maliciously, in which case the bucket may already exist and we're just re-establishing the connection and the bucket may already exist (aka try/except on bucket exists).

skurfer replied:
If a trail is deleted, it won’t be passed to the `process()` method, so we won’t have any metadata about it (like which bucket it was attached to). The bucket here is completely made up based on the name of the trail. (I suppose a collision is still possible, so we can test for existing.)

Should we allow (require?) a bucket name for each trail in the policy so the code doesn’t have to guess?

kapilt replied:
good point.. so the end goal is to have something like 'ensure' trail, such that no matter what an active trail is running, that we can trigger off events that modify or disable trails or for initial account setup.

kapilt commented on c7n/resources/cloudtrail.py:
```diff
@@ -31,3 +40,93 @@ class resource_type(object):
+            s3client.put_bucket_policy(Bucket=bucket_name, Policy=policy_json)
+            new_trail = client.create_trail(
+                Name=trail_name,
+                S3BucketName=bucket_name,
```
the other options here need to be exposed for config or defaulted (global events, multi-region, kms key, etc).

kapilt commented on c7n/resources/cloudtrail.py:
```diff
@@ -31,3 +40,93 @@ class resource_type(object):
+
+
+@actions.register('enable')
+class EnableTrail(BaseAction):
```
actions and filters need iam permissions as class attribute

skurfer commented:
This has been updated to support the various options.

Like I mentioned on the call, the coverage didn’t match my expectations. I ran through in the debugger and at least two of the lines that supposedly aren’t getting tested *are* getting called when tests run, so I’m chalking it up to weirdness in the coverage package and submitting this as is.

skurfer commented:
Updates based on the last discussion:

* Moved the enable trail action to the account resource
* Added an explicit call to `describe_trails()` since we’re now being passed a list of accounts instead of trails
* Added DescribeTrails to the list of permissions
* moved and updated tests

kapilt commented:
branch updated and pull request resubmitted in #1098
